test/test_fakeredis.py

```diff
@@ -25,4529 +21,4692 @@
 REDIS3 = REDIS_VERSION >= '3'
 
 
-UpdateCommand = namedtuple('UpdateCommand', 'input expected_return_value expected_state')
+redis2_only = pytest.mark.skipif(REDIS3, reason="Test is only applicable to redis-py 2.x")
+redis3_only = pytest.mark.skipif(not REDIS3, reason="Test is only applicable to redis-py 3.x")
+fake_only = pytest.mark.parametrize(
+    'create_redis',
+    [pytest.param('FakeStrictRedis', marks=pytest.mark.fake)],
+    indirect=True
+)
+
+
+def key_val_dict(size=100):
+    return {b'key:' + bytes([i]): b'val:' + bytes([i])
+            for i in range(size)}
+
+
+def round_str(x):
+    assert isinstance(x, bytes)
+    return round(float(x))
+
+
+def raw_command(r, *args):
+    """Like execute_command, but does not do command-specific response parsing"""
+    response_callbacks = r.response_callbacks
+    try:
+        r.response_callbacks = {}
+        return r.execute_command(*args)
+    finally:
+        r.response_callbacks = response_callbacks
+
+
+# Wrap some redis commands to abstract differences between redis-py 2 and 3.
+def zadd(r, key, d, *args, **kwargs):
+    if REDIS3:
+        return r.zadd(key, d, *args, **kwargs)
+    else:
+        return r.zadd(key, **d)
+
+
+def zincrby(r, key, amount, value):
+    if REDIS3:
+        return r.zincrby(key, amount, value)
+    else:
+        return r.zincrby(key, value, amount)
+
+
+@pytest.fixture(scope="session")
+def is_redis_running():
+    try:
+        r = redis.StrictRedis('localhost', port=6379)
+        r.ping()
+    except redis.ConnectionError:
+        return False
+    else:
+        return True
+
+
+@pytest.fixture(
+    params=[
+        pytest.param('StrictRedis', marks=pytest.mark.real),
+        pytest.param('FakeStrictRedis', marks=pytest.mark.fake)
+    ]
+)
+def create_redis(request):
+    name = request.param
+    if not name.startswith('Fake') and not request.getfixturevalue('is_redis_running'):
+        pytest.skip('Redis is not running')
+    decode_responses = request.node.get_closest_marker('decode_responses') is not None
+
+    def factory(db=0):
+        if name.startswith('Fake'):
+            fake_server = request.getfixturevalue('fake_server')
+            cls = getattr(fakeredis, name)
+            return cls(db=db, decode_responses=decode_responses, server=fake_server)
+        else:
+            cls = getattr(redis, name)
+            return cls('localhost', port=6379, db=db, decode_responses=decode_responses)
+
+    return factory
+
+
+@pytest.fixture
+def r(request, create_redis):
```

Comment:
Would the `r` fixture run before each test or is it cached somehow like `self.redis` used to be?

Reply:
It runs for each test. I don't think there was any caching before though - what caching are you referring to?

Reply:
Ah right, I was thinking of `self.redis` stored on the test object but that is recreated in `setup` before each test anyway. Forgot about that...